// latex-to-pdf/src/TemplateParser.ts
import * as fs from "fs";
import { TypeData } from "./ParseInterfaces";


export const bindingRegex = {
  binding: /{{let\s+(!)?([\w\d_]+):(\[)?(import|new|raw)<(.*?)(?:,"(.*?)")?>\]?}}/g,
};

export class TemplateParser {
  static extractFields(latexFilePath: string): TypeData {
    const content = fs.readFileSync(latexFilePath, "utf-8");

    const parsedData: TypeData = {
      bindings: [],
    };

    // Parse bindings
    for (const match of content.matchAll(bindingRegex.binding)) {
      const [,isField, ident, array, bindType, typeDetails, path] = match;
      parsedData.bindings.push({
        ident,
        array: !!array,
        objectType: typeDetails,
        type: bindType === "import"
          ? { type: "import", path: path || "" }
          : bindType === "new"
          ? { type: "new", schema: typeDetails.split(",").map(f => {
        const [name, type] = f.split(":");
        return { name, type };
          })
          }
          : { type: "raw", raw_type: typeDetails },
        isField: !isField,
      });
    }
    // Capture insertion points
    // const matchAndStore = (regex: RegExp, type: string) => {
    //   for (const match of content.matchAll(regex)) {
    //     parsedData.replacements.push({
    //       start: match.index!,
    //       end: match.index! + match[0].length,
    //       type,
    //       content: match[0], // Original LaTeX content
    //     });
    //   }
    // };

    // matchAndStore(regexPatterns.rawField, "rawField");
    // matchAndStore(regexPatterns.tableFunction, "tableFunction");
    // matchAndStore(regexPatterns.forLoop, "forLoop");
    // matchAndStore(regexPatterns.ifStatement, "ifStatement");

    return parsedData;
  }
}
