// latex-to-pdf/src/TypeGenerator.ts
import * as fs from "fs";
import * as path from "path";
import { TypeData } from "./ParseInterfaces";

export class TypeGenerator {
  static capitalizeFirstLetter(str: string): string {
    return str.charAt(0).toUpperCase() + str.slice(1);
  }

  static generateInterface(latexFilePath: string, typeData: TypeData): string {
    const fileName =
      this.capitalizeFirstLetter(
        path.basename(latexFilePath, path.extname(latexFilePath))
      ) + "LatexRenderData";
    const interfaceName = fileName.replace(/[^a-zA-Z0-9]/g, "_"); // Make it TypeScript-safe

    let output = "";

    for (const binding of typeData.bindings) {
      if (binding.type.type === "import") {
        output += `import type {${binding.objectType}} from "${binding.type.path}";\n`;
      }
    }

    output += `export interface ${interfaceName} {\n`;

    for (const binding of typeData.bindings) {
      if (binding.isField) {
        if (binding.type.type === "import") {
          output += `  ${binding.ident}: ${
            binding.array ? binding.objectType + "[]" : binding.objectType
          };\n`;
        } else if (binding.type.type === "new") {
          output += `  ${binding.ident}: ${
            binding.array ? binding.ident + "[]" : binding.ident
          };\n`;
        } else if (binding.type.type === "raw") {
          output += `  ${binding.ident}: ${binding.type.raw_type};\n`;
        }
      }
    }

    output += "}\n\n";

    // Generate any new object types
    for (const binding of typeData.bindings) {
      if (binding.type.type === "new") {
        output += `export interface ${binding.ident} {\n`;
        for (const field of binding.type.schema || []) {
          output += `  ${field.name}: ${field.type};\n`;
        }
        output += `}\n\n`;
      }
    }

    return output;
  }

  static writeInterface(
    latexFilePath: string,
    typeData: TypeData,
    outputDir: string = "./latex_interfaces"
  ): string {
    const interfaceContent = this.generateInterface(latexFilePath, typeData);
    const fileName = path.basename(latexFilePath, path.extname(latexFilePath));
    const outputPath = path.join(outputDir, `${fileName}.d.ts`);

    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    fs.writeFileSync(outputPath, interfaceContent);
    console.log(`✅ Type definitions written to ${outputPath}`);
    return fileName;
  }
}
