// latex-to-pdf/dist/TemplateParser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateParser = exports.bindingRegex = void 0;
const fs = __importStar(require("fs"));
exports.bindingRegex = {
    binding: /{{let\s+(!)?([\w\d_]+):(\[)?(import|new|raw)<(.*?)(?:,"(.*?)")?>\]?}}/g,
};
class TemplateParser {
    static extractFields(latexFilePath) {
        const content = fs.readFileSync(latexFilePath, "utf-8");
        const parsedData = {
            bindings: [],
        };
        // Parse bindings
        for (const match of content.matchAll(exports.bindingRegex.binding)) {
            const [, isField, ident, array, bindType, typeDetails, path] = match;
            parsedData.bindings.push({
                ident,
                array: !!array,
                objectType: typeDetails,
                type: bindType === "import"
                    ? { type: "import", path: path || "" }
                    : bindType === "new"
                        ? { type: "new", schema: typeDetails.split(",").map(f => {
                                const [name, type] = f.split(":");
                                return { name, type };
                            })
                        }
                        : { type: "raw", raw_type: typeDetails },
                isField: !isField,
            });
        }
        // Capture insertion points
        // const matchAndStore = (regex: RegExp, type: string) => {
        //   for (const match of content.matchAll(regex)) {
        //     parsedData.replacements.push({
        //       start: match.index!,
        //       end: match.index! + match[0].length,
        //       type,
        //       content: match[0], // Original LaTeX content
        //     });
        //   }
        // };
        // matchAndStore(regexPatterns.rawField, "rawField");
        // matchAndStore(regexPatterns.tableFunction, "tableFunction");
        // matchAndStore(regexPatterns.forLoop, "forLoop");
        // matchAndStore(regexPatterns.ifStatement, "ifStatement");
        return parsedData;
    }
}
exports.TemplateParser = TemplateParser;
